Generating event lists and 2 IRFs, to create DL3 fits file

chaimain opened the pull request:
Proceeding with creating event lists of LST DL2 files as per the standard https://gamma-astro-data-formats.readthedocs.io/en/latest/ and using pyIRF to create simple point-like IRFs - Energy Dispersion and Effective Area.

Usage of pyIRF is without using any cuts optimization for now, and have made a copy of the config file used by pyIRF.
pyIRF's master branch still has old dependencies (gammapy and ctapipe) which were not compatible for the current script. I used a separate [branch](https://github.com/chaimain/pyirf/tree/gammapy017) with gammapy v0.17 dependency to produce the IRFs.

I have made a separate file for cleaning the data in a separate new path lstchain/clean/. This is most likely unnecessary, but I tried to have a separate code instead of the filter_events in lstchain/reco/utils.py, to separate it with user-defined cuts.

I have also made a separate file hdu_table.py in a separate new path lstchain/hdu/ to include the main functions required in creating the event lists and HDU and Obs index files.

The event list script works on a single file right now, but can easily be changed to work for a list of runs in the same path. Accordingly, I separated the script for creating HDU and Obs index fits files from a list of DL3 fits files, as I could only write it once or overwrite the existing.

MaxNoe commented:
pyirf has no released version yet and will change very much before and during the coding sprint at the end of september , so I wouldn't really try to use it before that is over and a first release is made.

chaimain commented:
Yes, I understand that and would look into incorporating the updated version of pyIRF to create the IRFs for the DL3. I have added the option to not include the IRFs and just create DL3 files with the event lists for now. The IRFs generated are very basic and so it seemed ok to just try to include them with the event lists to generate DL3 file with LST data.
The draft PR is also intended to understand how the structure of generating DL3 files from DL2 files can be in the future.

chaimain commented:
I have updated the core pyIRF functions with the current developer version of pyIRF (v0.3.0dev0), changed the selection cuts config file from .yml to .json and added some tests .
There are still things to be updated, most of which shall be done in a few days.

MaxNoe commented on lstchain/hdu/hdu_table.py:
```diff
@@ -0,0 +1,464 @@
+#once the pyIRF IRF HDUs are generalized for our use, we can import the functions
+# directly from pyIRF
+
+def make_aeff2d_hdu(
```
I do not understand why you duplicated these functions. They are besides the adding of an additional dimesion to the effecive area exactly the same as in pyirf.

You should add the additional dimension and then just call the pyirf function.

Something like this
```
def add_dummy_area_bin(effective_area):
    return np.column_stack([effective_area, np.zeros_like(effective_area)])

```

And then call the function from pyirf:

```
effective_area = add_dummy_area_bin(effective_area_per_energy(...))

hdu = create_effective_area_hdu(...)
```

Duplicating the code does not make sense

chaimain replied:
Ah! Yes, I can do this. I assumed that as I am using gammapy to analyse the DL3 files, and where they still have the issue you mentioned, it will be safer to use the recreate it here, until it is resolved both in pyirf and gammapy. But I understand your point and will make the changes

MaxNoe commented on lstchain/hdu/hdu_table.py:
```diff
@@ -0,0 +1,464 @@
+
+    return fits.BinTableHDU(aeff, header=header, name=extname)
+
+def make_edisp2d_hdu(
```
Again, why did you copy this? The function of pyirf is the same.

MaxNoe commented on lstchain/hdu/hdu_table.py:
```diff
@@ -0,0 +1,464 @@
+    filename_obs_table = 'obs-index.fits.gz'
+    obs = fits.BinTableHDU(obs_table)
+    hdu_list = fits.HDUList()
+    hdu_list.append(obs)
```
You need a primary HDU as the first hdu, just do `hdu_list= fits.HDUList([fits.PrimaryHDU] + obs)`

chaimain replied:
Understood

MaxNoe commented on lstchain/hdu/hdu_table.py:
```diff
@@ -0,0 +1,464 @@
+
+    ##########################################################################
+    ### Event table columns
+    event_table["EVENT_ID"] = Column(
```
It's much easier and much less code to just use an astropy table than declaring these low-level fits objects.

vuillaut commented on environment.yml:
```diff
@@ -6,6 +6,7 @@ channels:
 dependencies:
   - python=3.7
   - pip
+  - pyirf>=0.2.0
```
I think there will be a release soon and this should be changed to the latest version

MaxNoe replied:
And requirung == probably makes sense. Or at least `!=0.x.0` as there will probably be breaking changes down the road until we have a more stable interface.

maxnoe replied:
Change to `~=0.3.0`, we made the release today.

vuillaut commented on lstchain/clean/data_clean.py:
```diff
@@ -0,0 +1,40 @@
+
+    return data
+
+def data_filter(data):
```
such function already exists
`lstchain.io.io.filter_events`

vuillaut replied:
You might want a more specific function for your use case but you can reuse `filter_events`

chaimain replied:
I thought that using filter_events and applying cuts, might be slightly different, hence tried this way. But I will try with filter_events and check it.

vuillaut replied:
simple suggestion, if it does not fit your needs, make your function no problem (it would just be less code)

MaxNoe replied:
Filtering events is applying cuts. No semantic difference.

vuillaut reviewed (changes requested):
I think it would be appropriate to create an `irf` submodule now also.

vuillaut commented on lstchain/data/data_selection_cuts.json:
```diff
@@ -0,0 +1,15 @@
+    "intensity": 200,
+    "r": 1,
+    "wl": 0.1,
+    "leakage2_intensity": 0.2,
```
does this parameter still exist?

chaimain replied:
To use it for pre v0.6 format files. I have created a condition in the data_clean.py script to account for it.

MaxNoe replied:
No, it is the old name of the parameter one line below

MaxNoe commented on lstchain/clean/data_clean.py:
```diff
@@ -0,0 +1,40 @@
+    data = data[data['wl'] > data_cut["general"]["wl"]]
+    data = data[data['gh_score'] > data_cut["general"]["gh_score"]]
+    #data = data[data['theta']] < data_cut["mc"]["theta_cut"]
+    data = data[data['source_fov_offset'].value < data_cut["general"]["source_fov_offset"]]
```
Never use `.value` when unsure of units. Use `to_value(unit)` to be sure to get a numpy array with the correct unit.`

chaimain replied:
OK, understood.

MaxNoe commented on lstchain/data/data_selection_cuts.json:
```diff
@@ -0,0 +1,15 @@
+    "alpha": 8
+    },
+  "mc":{
+    "theta_cut": 0.14
```
I would highly recommend to also use units in the config file. This could be achieved like this:

```
"theta_cut": {"value": 0.14, "unit": "deg"}
```

And then reading with:

```
theta_cut = u.Quantity(**config["theta_cut"])
```

chaimain replied:
Ok thanks! I will make the change.

chaimain replied:
Using units in this config file, would not let me use filter_events as it is. Would you then recommend to make this config file like the lstchain_standard_config to suit filter_events? As only the angular_offset parameters need units, so those cuts can be just adjusted in the main script itself.

maxnoe commented on lstchain/clean/data_clean.py:
```diff
@@ -0,0 +1,40 @@
+    data = data[data['intensity'] > data_cut["general"]["intensity"]]
+    data = data[data['wl'] > data_cut["general"]["wl"]]
+    data = data[data['gh_score'] > data_cut["general"]["gh_score"]]
+    data = data[data['theta'].to_value() < data_cut["mc"]["theta_cut"]]
```
`to_value` needs a `unit` argument: `to_value(u.deg)` for example.

maxnoe replied:
But again, I would also put the `theta_cut` as unit, and then you don't need the `to_value` at all.

maxnoe commented on lstchain/clean/data_clean.py:
```diff
@@ -0,0 +1,40 @@
+    data_cut = json.load(json_file)
+
+### TODO: We can generalize for the user to choose which selection cuts to make
+def mc_filter(data):
```
These filtering functions are the same? And they have to be the same for data and mc.

chaimain replied:
Yes, they will be. Just that the theta² cut is not applied on real data now. It will be used in gammapy analysis, when selecting the dataset's on_region. Also, I am going to remove this clean submodule and just use filter_events and separate fixed cuts of g/h score and theta cuts, directly in the main script.

maxnoe commented on lstchain/hdu/hdu_table.py:
```diff
@@ -0,0 +1,350 @@
+        directory containing the fits file
+    """
+
+    # empty lists with the quantities we want
```
Making a list for each of the required variables is very verbose and makes it hard to follow.

I would recommend to use a single list of dictionaries instead, which makes it most of the time much easier to follow and a lot less code, which is also easier to maintain, because adding something only requires change at a single position, not at two.

maxnoe commented on lstchain/hdu/hdu_table.py:
```diff
@@ -0,0 +1,350 @@
+    # loop through the files
+    for file in filename_list:
+
+        if os.path.exists(str(fits_dir)+"/"+file):
```
Don't use string concatenation for paths. Use `os.path.join` or `pathlib.Path` operations.

maxnoe replied:
Why do you need to convert fits_dir to `str` anyways? If it is already `pathlib.Path`, then simply use `(fits_dir / file).is_file()`

chaimain replied:
I will change it accordingly, thanks!

maxnoe commented on lstchain/hdu/hdu_table.py:
```diff
@@ -0,0 +1,350 @@
+
+        if os.path.exists(str(fits_dir)+"/"+file):
+            try:
+                event_table = Table.read(str(fits_dir)+"/"+file, hdu="EVENTS")
```
same as above

maxnoe commented on lstchain/hdu/hdu_table.py:
```diff
@@ -0,0 +1,350 @@
+                gti_table = Table.read(str(fits_dir)+"/"+file, hdu="GTI")
+                pointing_table = Table.read(str(fits_dir)+"/"+file, hdu="POINTING")
+            except Exception:
+                print(f"fits corrupted for file {file}")
```
Broken data files should not be ignored I think. This should raise an exception!

Also, use logging, not print statements.

chaimain replied:
Yes, I had forgotten about this. Will amend it.

maxnoe commented on lstchain/hdu/hdu_table.py:
```diff
@@ -0,0 +1,350 @@
+                print(f"fits corrupted for file {file}")
+                continue
+        else:
+            print(f"fits {file} doesn't exist")
```
This should also raise an error. The user wanted a file to be included, if it is not there, ignoring that silently is not good.

maxnoe replied:
Where silently is relative, I know, but I think just printing it is not enough

maxnoe commented on lstchain/hdu/hdu_table.py:
```diff
@@ -0,0 +1,350 @@
+
+        ###############################################
+        #Energy Dispersion
+        if Table.read(str(fits_dir)+"/"+file, hdu="EDISP"):
```
same as above, don't use str operations here.

chaimain commented:
I am closing this draft PR, as this was a mess with some erroneously added 'git pulls' from different branches. I will make a new draft PR, with incorporating all the comments mentioned here. I apologize again for the mess and repetition.
